modelzoo/ESMM/train.py

```diff
@@ -0,0 +1,882 @@
+                                                       dimension=16,
+                                                       combiner='mean')
+
+
+        if column_name in USER_COLUMN:
+            user_column.append(embedding_column)
+        elif column_name in ITEM_COLUMN:
+            item_column.append(embedding_column)
+        elif column_name in COMBO_COLUMN:
+            combo_column.append(embedding_column)
+
+    return user_column, item_column, combo_column
+
+def train(sess_config,
+          input_hooks,
+          model,
+          train_init_op,
+          train_steps,
+          keep_checkpoint_max,
+          checkpoint_dir,
+          save_steps=None,
+          timeline_steps=None,
+          no_eval=None,
+          tf_config=None,
+          server=None,
+          stock_tf=None,
+          incremental_ckpt=None):
+    hooks = []
+    hooks.extend(input_hooks)
+
+    scaffold = tf.train.Scaffold(
+        local_init_op=tf.group(tf.local_variables_initializer(), train_init_op),
+        saver=tf.train.Saver(max_to_keep=keep_checkpoint_max))
+
+    stop_hook = tf.train.StopAtStepHook(last_step=train_steps)
+    log_hook = tf.train.LoggingTensorHook(
+        {'steps': model.global_step,
+         'loss': model.loss
+        }, every_n_iter=100)
+    hooks.append(stop_hook)
+    hooks.append(log_hook)
+
+    if timeline_steps and timeline_steps > 0:
+        hooks.append(tf.train.ProfilerHook(save_steps=timeline_steps,
+                                           output_dir=checkpoint_dir))
+    save_ckp_steps = save_steps if save_steps or no_eval else train_steps
+
+    '''
+                            Incremental_Checkpoint
+    To enable Incremental_Checkpoint please `save_incremental_checkpoint_secs`
+    in 'tf.train.MonitoredTrainingSession' checkpoint_dir must be defined. By default
+    save_incremental_checkpoint_secs is None. Incremental_save checkpoint time
+    in seconds can be set to use incremental checkpoint function, like
+    `tf.train.MonitoredTrainingSession(save_incremental_checkpoint_secs=args.incremental_ckpt)`
+    '''
+    if not stock_tf and incremental_ckpt:
+        raise ValueError('Incremental_Checkpoint has not been enabled yet.' \
+                         'Please see comments in the code.')
+        # with tf.train.MonitoredTrainingSession(
+        #         master=server.target if server else '',
+        #         is_chief=tf_config['is_chief'] if tf_config else True,
+        #         hooks=hooks,
+        #         scaffold=scaffold,
+        #         checkpoint_dir=checkpoint_dir,
+        #         save_checkpoint_steps=save_ckp_steps,
+        #         summary_dir=checkpoint_dir,
+        #         save_summaries_steps=save_steps,
+        #         config=sess_config,
+        #         save_incremental_checkpoint_secs=incremental_ckpt) as sess:
+        #     while not sess.should_stop():
+        #         sess.run([model.loss, model.train_op])
```

Comment:
Because Incremental_Checkpoint is disabled by default and it needs an extra dozen lines of code under an 'if' block to enable, so I use a comment in WDL. Use Error and remove the commented code. Or remove Error, uncomment these codes, and edit comments.

Reply:
Ok. I removed the code and stay with the error. But could you elaborate more on Incremental_Checkpoint? I tried implementing it with hints from https://deeprec.readthedocs.io/zh/latest/Incremental-Checkpoint.html . Is something else required in model's implementations to make it work or is this feature (Incremental_Checkpoint) not enabled in DeepRec yet?

Reply:
I think your previous commented code is ok. Did any error occur? If some error occurs, please show me the log.

Reply:
No error occured. I'm referring to raising a error in train.py line 458. You said: 

> Use Error and remove the commented code. Or remove Error, uncomment these codes, and edit comments.

So I used Error and removed the commented code.